server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -113,12 +118,21 @@ public DesiredBalance compute(
             logger.debug("Recomputing desired balance for [{}]", desiredBalanceInput.index());
         }
 
+        // A new ClusterInfo has arrived, clear the tracking for started shards
+        if (lastClusterInfo != desiredBalanceInput.routingAllocation().clusterInfo()) {
+            lastClusterInfo = desiredBalanceInput.routingAllocation().clusterInfo();
+            shardsStartedByAllocate = new HashMap<>();
+        }
```

Comment:
I think one issue with this approach is that a shard allocated in last polling of ClusterInfo but started in this polling cycle is not accounted for, i.e. the following sequence of events:
1. 1st CluserInfo poll
2. Allocator moves a shard, but it has not started on the target node
3. 2nd ClusterInfo poll
4. The shard starts on the target node

I think we want the started shard contributing to the simulations performed in the 2nd ClusterInfo polling cycle?

Reply:
Would that not be simulated by the shard started done in the beginning of compute?